Running over large/hard objects in vehicles will damage your wheels

RenechCDDA opened the pull request:
#### Summary
Balance "Running over large/hard objects in vehicles will damage your wheels"

#### Purpose of change
Just something that's been missing

#### Describe the solution
Add it

**(Note: The below description is no longer accurate, see conversations in the comments below)**

```
    // damage is calculated based on the size of roadkill relative to the wheel's volume and hardness of the roadkill
    // so e.g. a single nail is not likely to damage a huge tire by itself, but a whole road full of them almost certainly will.
    // While a road full of feathers just gets flattened, because they're the softest thing possible! Even if there's a lot of them!
```

The calculation compares ( the material hardness of the roadkill item vs the material hardness of the wheel item) * (volume of the roadkill item vs the volume of the wheel item)... neither factor can exceed 100% contribution. The result is that huge wheels made of stronger materials (e.g. steel) are fundamentally more difficult to damage than tiny wheels. An office chair with tiny casters running over a chunk of steel can easily damage it, but a road roller is very likely to shrug off a single chunk of steel. Similarly an armored wheel is harder to damage than a commercial one made of rubber.

The important thing to note is that relative hardness is specifically scaled to be exponentially more important - even a single nail can ruin a tire's day. (Though the calculations are *very* generous for how likely that is)

#### Describe alternatives you've considered
Having it damage other vehicle parts besides just wheels

#### Testing
Autodrive avoids running over items:

<img width="1593" height="1028" alt="image" src="https://github.com/user-attachments/assets/9690563d-a858-4474-9c0d-9b4880c69a3a" />

Some test data/examples of running stuff over:
https://github.com/CleverRaven/Cataclysm-DDA/pull/83004#issuecomment-3324396627


#### Additional context

Known issues:
It only rolls vs top level items so a canvas bag full of nails is just treated like a really big, canvas bag.

GuardianDll reviewed (changes requested):
I agree the way for player to  know where it is safe to drive is needed - tileset can easily hide a giant pile of corpses or stones behind the random texture of a small item, and curses do not have any

GuardianDll commented:
I would still prefer a visible indicator of tile fullness or sort of, because sometimes you cannot rely on autodrive (like when you need to run like hell, pedal to the metal), but if autodrive intend to handle it for casual driving, i am okay with that

PatrikLundell commented:
I remain adamantly against this. Autodrive being capable to handle it at a crawl is useless when you're trying to drive in the worst areas (cities), because you need to be on the lookout for monsters, and so can't use it there (apart from it being unable to drive there unless you've cleared paths for it), and outside of those areas it's just too slow to get anywhere.

RenechCDDA commented:
I've added some test cases, so we can see how three separate wheels will perform vs a variety of items. The wheels tested are bicycle wheel (light), regular wheel (medium), and armored wheel (heavy). The items tested against are a tempered steel chunk, a body pillow, a fake corpse, an **enormous** exodii steel ingot, and a watermelon.

I'm not terribly happy with the distribution of chances right now, volume is still much too large of a factor. Additionally the chances of a huge steel ingot vs a normal car wheel are significantly below 100%, and hardness is actually *why* it's less than 100%.

I think I may switch to calculating versus the **weakest** material for the wheel. So a regular wheel would be treated as if it is 100% rubber, as opposed to 50% rubber/50% steel. (Alternatively we could maybe just update the wheels' material definitions??)

ShnitzelX2 commented on src/vehicle_autodrive.cpp:
```diff
@@ -772,6 +772,11 @@ bool vehicle::autodrive_controller::check_drivable( map &here, const tripoint_bu
         return false;
     }
 
+    // don't drive over junk which could damage our wheels, the player must always manually do that.
+    if( here.has_items( pt ) ) {
+        return false;
+    }
+
```
> (the entire vehicle, not just the wheels!)

I'm not recommending this as part of your design, it was just a remark about the pathing. 

> there is no item which is 100% safe to run over

That's how the math works out now, but I'd prefer a threshold that would make items with, let's say, less than 1% chance to damage tires always have a 0% chance to damage tires. Examples: a piece of paper, leaves, a plastic bag, a twig, small mostly-harmless things. Then autodrive could ignore those items.

I'd definitely make some test cases first to see different item chances, though.

kevingranade commented:
After some quick searching around I'm seeing a lot of anecdotal statements from both individuals and organizations saying roadkill presents a meaningful risk of puncture.
https://www.google.com/search?q=tire+puncture+from+roadkill
A rare case where anecdotal information "here's a picture of a bone sticking out of my tire" is PLENTY of evidence.

As for actual implementation  I wonfer if individual items need to be tagged because I still don't think there are that many of them, and I don't know that trying to determine it based on material and gross shape stats is going to do it.

GuardianDll commented:
Cars ignoring the items on ground akile the character and monsters do implies not only that it should not be applied to cars, but also that it should apply to both cars and creatures walking around
>  tires aren't very well represented to begin with

Can you elaborate?

RenechCDDA commented:
Okay, so basically what needs to happen here is:
-Switch to only checking for damage vs items with some specific flag/enumerated property
-Absolutely remove volume from the calculations
-Make the hardness vs hardness calculations not... bad, or otherwise make sure the results make sense

RenechCDDA committed:
Running over large/hard objects in vehicles will damage your wheels

RenechCDDA committed:
Refactor damage chance vs wheels

-Completely remove volume from the calculation
-Only items with a specific flag can damage wheels
-Wheels can have a special flag which increases their hardness
-Greatly simplify test execution
-Enforce lower numeric bounds of 0 for chip_resist. (Wheel damage chance absolutely *does not work* with negative numbers there)

Eiritha commented:
> Concerns: -Damage chance for something _cotton_ vs a rubber wheel is very very high, I would expect it to be 0 or nearly 0

This is very unlikely to be new information and I apologize if it's unhelpful (or wrong- I am not necessarily good at reading code or understanding values), but so far as I can tell this is happening because cotton has a chip_resist of 6 compared to, say, the 2 of flesh, leading to the rather unusual circumstance in which the T-shirt and track pants combo worn by a recently deceased Deadite would be significantly more destructive to a car's wheel than the corpse itself as the numbers stand.

This is of course hilarious, and I firmly approve of the mental image of deploying a line of Levi Ackerman body pillows to disable military convoys.  It may not, however, lead to the desired game outcome.

Termineitor244 commented:
Great! About the flags, my only wish would be to have an additional `Magical_Wheel` flag (or something) for wheels in Magiclysm made of mana, or I guess for levitating vehicles too (thought those really need a whole new system), that allows a wheel to dismiss the chance of being damaged when running over something, since they are not made of rubber, but a material that should ignore those kinds of smalls points of rupture as long as you supply them with mana.

RenechCDDA commented:
> Great! About the flags, my only wish would be to have an additional `Magical_Wheel` flag (or something) for wheels in Magiclysm made of mana, or I guess for levitating vehicles too (thought those really need a whole new system), that allows a wheel to dismiss the chance of being damaged when running over something, since they are not made of rubber, but a material that should ignore those kinds of smalls points of rupture as long as you supply them with mana.

Do we have any wheels that work that way in Magiclysm right now?


Oops, noticed that the greeter bot is malfunctioning. Sorry about that, Eiritha.

kevingranade commented:
If that motorcycle should be immune to all damage I feel like that's an "immune to all damage" flag, not a specific thing for wheels?  If not, why would it be wheels in particular that would get it?

RenechCDDA committed:
Add DAMAGE_VEHICLE_WHEELS flag to a bunch of items

RenechCDDA committed:
Switch vehicle runover tests to use test iems

Remove flags from inappropriate vanilla items
